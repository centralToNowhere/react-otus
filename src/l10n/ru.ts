export const l10n = {
  maxWidthLabel: "Макс. ширина:",
  maxHeightLabel: "Макс. высота:",
  cellSizeLabel: "Размер ячейки:",
  capacityLabel: "Процент заполненности:",
  speedLabel: "Обновлений в секунду:",
  buttonStart: "Старт",
  buttonStop: "Стоп",
  buttonReset: "Сброс",
  buttonGenerateRandom: "Создать",

  // Экран регистрации
  gameNameFull: "Conway's Game of Life",
  gameHeadingPart1: "Conway's",
  gameHeadingPart2: "Game of Life",
  buttonStartGameAsPlayer: "Начать игру",
  rulesHeading: "Правила",
  description1: "Игра «Жизнь» (англ. ",
  source: "Источник:",
  sourceName: "Игра «Жизнь»",
  description2:
    ") — клеточный автомат, придуманный английским математиком Джоном Конвеем в 1970 году.",
  rule1:
    "Место действия этой игры — «вселенная» — это размеченная на клетки поверхность или плоскость — безграничная, ограниченная, или замкнутая (в пределе — бесконечная плоскость).",
  rule2:
    "Каждая клетка на этой поверхности может находиться в двух состояниях: быть «живой» (заполненной) или быть «мёртвой» (пустой). Клетка имеет восемь соседей, окружающих её.",
  rule3:
    "Распределение живых клеток в начале игры называется первым поколением. Каждое следующее поколение рассчитывается на основе предыдущего по таким правилам:",
  rule3_1:
    "в пустой (мёртвой) клетке, рядом с которой ровно три живые клетки, зарождается жизнь",
  rule3_2:
    "если у живой клетки есть две или три живые соседки, то эта клетка продолжает жить; в противном случае, если соседей меньше двух или больше трёх, клетка умирает («от одиночества» или «от перенаселённости»)",
  rule4: "Игра прекращается, если:",
  rule4_1: "на поле не останется ни одной «живой» клетки",
  rule4_2:
    "конфигурация на очередном шаге в точности (без сдвигов и поворотов) повторит себя же на одном из более ранних шагов (складывается периодическая конфигурация)",
  rule4_3:
    "при очередном шаге ни одна из клеток не меняет своего состояния (складывается стабильная конфигурация; предыдущее правило, вырожденное до одного шага назад)",

  // Контейнер игрока
  registerPlayerLabel: "Введите имя игрока",
  logoutButton: "Выйти",
  generation: "Поколение",

  // Ошибки
  minCellsAmount: "Размер игрового поля меньше размера ячейки.",
  maxCellsAmount: "Превышено максимальное количество ячеек",
  positiveNumber: "Ожидается положительное число.",
  nonNegativeNumber: "Ожидается неотрицательное число.",
};
