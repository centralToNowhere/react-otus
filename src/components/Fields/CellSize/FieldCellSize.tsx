import React, { useCallback, useState } from "react";
import { l10n } from "@/l10n/ru";
import {
  InputField,
  LabelField,
  onBlurHandler,
  useOnChangeHandler,
} from "@/components/Fields";
import { FormField } from "@/components/Form/FormField";
import { InputPatterns } from "@/components/Fields";
import { FieldError } from "@/components/Fields/FieldError/FieldError";
import { Form, IFieldProps } from "@/components/Form";
import { useDebounce, isValidCellSizeString } from "@/utils";
import {
  onDirtyBlurHandler,
  onDirtyChangeHandler,
} from "@/components/Fields/FieldHandlers";

export const FieldCellSize: React.FC<IFieldProps> = (props) => {
  const [cellSizeString, setCellSizeString] = useState<string>(props.value);
  const [error, setError] = useState({
    show: false,
    msg: "Expected number >= 10",
  });

  const validateCellSize = useCallback((value: unknown): boolean => {
    return isValidCellSizeString(value);
  }, []);

  const onChangeDebounced = useDebounce<string>(
    useOnChangeHandler(
      props.onChange,
      validateCellSize,
      setCellSizeString,
      setError
    ),
    Form.inputDelay
  );

  const onChange = onDirtyChangeHandler((value) => {
    setCellSizeString(value);
    onChangeDebounced(value);
  });

  const onBlur = onDirtyBlurHandler((value) => {
    onChangeDebounced.clear();
    onBlurHandler(props.onChange, validateCellSize, setError)(value);
  });

  return (
    <FormField>
      <LabelField htmlFor="cell-size">{l10n.cellSizeLabel}</LabelField>
      <InputField
        id="cell-size"
        type="number"
        pattern={InputPatterns.number}
        step="1"
        name="cellSize"
        value={cellSizeString}
        autoComplete="off"
        onChange={onChange}
        onBlur={onBlur}
      />
      <FieldError show={error.show} msg={error.msg} />
    </FormField>
  );
};
