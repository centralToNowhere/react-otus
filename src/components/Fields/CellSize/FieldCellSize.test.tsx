import React from "react";
import { render, screen } from "@testing-library/react";
import { FieldCellSize } from "./FieldCellSize";
import { l10n } from "@/l10n/ru";

describe("FieldCellSize tests", () => {
  it("should render cell size input", () => {
    const { asFragment } = render(
      <FieldCellSize
        value={"40"}
        onChange={() => {
          // empty function
        }}
        onBlur={() => {
          // empty function
        }}
        error={{
          show: false,
          msg: "Expected number >= 10",
        }}
      />
    );

    const input: HTMLInputElement = screen.getByLabelText(l10n.cellSizeLabel);

    expect(input).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
