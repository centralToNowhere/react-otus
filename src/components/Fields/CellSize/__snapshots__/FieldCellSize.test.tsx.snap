// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldCellSize tests should render cell size input 1`] = `
<DocumentFragment>
  <div
    class="css-mkf0o6"
  >
    <label
      class="css-1wnvjz2"
      for="cell-size"
    >
      Размер ячейки:
    </label>
    <input
      autocomplete="off"
      class="css-16tc96n"
      id="cell-size"
      name="cellSize"
      pattern="[0-9]*"
      step="1"
      type="number"
      value="40"
    />
  </div>
</DocumentFragment>
`;
